nupic/research/frameworks/vernon/mixins/dendrite_metrics.py

```diff
@@ -0,0 +1,247 @@
+            plot_freq = plotting_args.get("plot_freq", 1)
+            plot_args = plotting_args.get("plot_args", {})
+            max_samples_to_plot = plotting_args.get("max_samples_to_plot", 1000)
+
+            assert callable(plot_func)
+            assert isinstance(plot_freq, int)
+            assert isinstance(max_samples_to_plot, int)
+            assert plot_freq > 1
+            assert max_samples_to_plot > 0
+
+            # The arguments may be given as a callable; useful for sampling random
+            # values that dictate plotting to, say, only plot a subset of units
+            if callable(plot_args):
+                plot_args = plot_args()
+            assert isinstance(plot_args, dict)
+
+            new_metric_args[metric_name] = dict(
+                plot_func=plot_func,
+                plot_freq=plot_freq,
+                plot_args=plot_args,
+                max_samples_to_plot=max_samples_to_plot,
+            )
+
+            all_max_num_samples.append(max_samples_to_plot)
+
+        max_samples_to_plot = max(all_max_num_samples)
+        return new_metric_args, filter_args, max_samples_to_plot
+
+    def run_epoch(self):
+        """
+        This runs the epoch with the hooks in tracking mode. The resulting 'activations'
+        and 'winning_masks' collected by these hooks are plotted via each 'plot_func'
+        along with their corresponding targets.
+        """
+
+        # Run the epoch with tracking enabled.
+        with self.dendrite_hooks:
+            results = super().run_epoch()
+
+        # The epoch was iterated in `run_epoch` so epoch 0 is really epoch 1 here.
+        iteration = self.current_epoch - 1
+
+        # Gather and plot the statistics.
+        for name, _, activations, winners in self.dendrite_hooks.get_statistics():
+
+            # Keep track of whether a plot is made below. If so, save the raw data.
+            plot_made = False
+
+            # Each 'plot_func' will be applied to each module being tracked.
+            for metric_name, plotting_args in self.metric_args.items():
+
+                # All of the defaults were set in `process_args`.
+                plot_func = plotting_args["plot_func"]
+                plot_freq = plotting_args["plot_freq"]
+                plot_args = plotting_args["plot_args"]
+                max_samples_to_plot = plotting_args["max_samples_to_plot"]
+
+                if iteration % plot_freq != 0:
+                    continue
+
+                # Only use up the the max number of samples for plotting.
+                targets = self.targets[:max_samples_to_plot]
+                activations = activations[:max_samples_to_plot]
+                winners = winners[:max_samples_to_plot]
+
+                # Call and log the results of the plot function.
+                # Here, "{name}" is the name of the module.
+                visual = plot_func(activations, winners, targets, **plot_args)
```

Comment:
This will cause an error if `plot_func` doesn't accept all 3 of 'activations', 'winners', 'targets' as inputs - which is the case for some plotting functions.

Reply:
In the example config, you'll notice it defines
```
def _plot_entropy_distribution(_dendrite_activations, winning_mask, targets):
    """
    This adjusts the function's signature to align with the mixin.
    """
    return plot_entropy_distribution(winning_mask, targets)
```

So whatever functions we define, we'll adjust the signature in a similar way and just take a dummy argument. I thought this was the most flexible way forward.